File: appwire/boot_generation.go
package appwire

import (
	"strconv"
	"strings"
)

// DaemonlessBootGeneration is the boot generation the hub stamps on a read it
// serves from the transcript with no daemon running for the session.
const DaemonlessBootGeneration = "daemonless"

// BootGeneration spells a daemon's boot counter as the wire carries it for
// the daemon's root thread.
func BootGeneration(counter uint64) string {
	return strconv.FormatUint(counter, 10)
}

// DescendantBootGeneration is the boot generation of a descendant served by
// its root's daemon: "<n>@<rootSessionID>", the root's generation qualified
// by the root. A descendant has no counter of its own, so its generation is
// comparable only with others from the same root's daemon; a daemon serving
// the descendant as its own root carries a plain counter, which replaces.
func DescendantBootGeneration(rootGeneration, rootSessionID string) string {
	return rootGeneration + "@" + rootSessionID
}

// BootGenerationAction is what a client does with a read response or
// history update, given the boot generation it holds for the thread.
type BootGenerationAction int

const (
	// BootGenerationApply: the same token as held; apply normally.
	BootGenerationApply BootGenerationAction = iota
	// BootGenerationIgnore: a lower generation of the same counter (the same
	// root, qualified or not), from a daemon boot the client already moved
	// past.
	BootGenerationIgnore
	// BootGenerationReplace: any other token (a higher number, another
	// counter, or a switch between numeric and daemonless). The client marks
	// the thread invalid, drops updates, and replaces the whole history with
	// a fresh subscribing latest-window read.
	BootGenerationReplace
)

// CompareBootGeneration is the spec's one state machine for boot generations
// (docs/superpowers/specs/2026-09-25-transcript-read-model-design.md,
// "Recorded length, entry ordinals and Seq"): it decides what a client
// holding held does with a message carrying incoming. Two tokens compare
// numerically only when both are counters of the same kind: both plain, or
// both qualified by the same root (DescendantBootGeneration). Anything else
// that differs replaces.
func CompareBootGeneration(held, incoming string) BootGenerationAction {
	if held == incoming {
		return BootGenerationApply
	}
	heldCounter, heldOwner, heldOK := parseBootGeneration(held)
	incomingCounter, incomingOwner, incomingOK := parseBootGeneration(incoming)
	if !heldOK || !incomingOK || heldOwner != incomingOwner {
		return BootGenerationReplace
	}
	switch {
	case incomingCounter < heldCounter:
		return BootGenerationIgnore
	case incomingCounter == heldCounter:
		return BootGenerationApply
	default:
		return BootGenerationReplace
	}
}

// parseBootGeneration reads a counter token: "<n>" (owner "") or
// "<n>@<root>". daemonless and malformed tokens report false.
func parseBootGeneration(token string) (counter uint64, owner string, ok bool) {
	number, owner, qualified := strings.Cut(token, "@")
	if qualified && owner == "" {
		return 0, "", false
	}
	counter, err := strconv.ParseUint(number, 10, 64)
	return counter, owner, err == nil
}

File: appwire-client/typescript/bootGeneration.ts
// The client half of the boot generation state machine
// (docs/superpowers/specs/2026-09-25-transcript-read-model-design.md,
// "Recorded length, entry ordinals and Seq"). It mirrors
// appwire.CompareBootGeneration (appwire/boot_generation.go) exactly, and its
// test mirrors the Go table row for row.

/** The boot generation the hub stamps on a read it serves from the transcript
 * with no daemon running for the session (appwire.DaemonlessBootGeneration). */
export const DAEMONLESS_BOOT_GENERATION = "daemonless";

/**
 * What a client does with a read response or history update, given the boot
 * generation it holds for the thread:
 * - `apply`: the same token as held; apply normally.
 * - `ignore`: a lower generation of the same counter, from a daemon boot the
 *   client already moved past.
 * - `replace`: any other token (a higher number, another counter's token, a
 *   switch between numeric and daemonless, or nothing held yet). The thread's
 *   whole history is replaced by a fresh latest-window read.
 */
export type BootGenerationAction = "apply" | "ignore" | "replace";

/**
 * Compares boot generation tokens: "<n>" (a daemon's own session),
 * "<n>@<rootSessionID>" (a descendant served by its root's daemon) or
 * "daemonless". Two tokens compare numerically only when both are counters of
 * the same owner; anything else that differs replaces.
 */
export function compareBootGeneration(held: string, incoming: string): BootGenerationAction {
  if (held === incoming) return "apply";
  const heldCounter = parseBootGeneration(held);
  const incomingCounter = parseBootGeneration(incoming);
  if (!heldCounter || !incomingCounter || heldCounter.owner !== incomingCounter.owner) return "replace";
  if (incomingCounter.counter < heldCounter.counter) return "ignore";
  if (incomingCounter.counter === heldCounter.counter) return "apply";
  return "replace";
}

const MAX_UINT64 = (1n << 64n) - 1n;

// Reads a counter token the way Go's strconv.ParseUint does (decimal digits
// only, within uint64), as a bigint so counters past 2^53 still compare
// exactly. daemonless and malformed tokens read as undefined.
function parseBootGeneration(token: string): { counter: bigint; owner: string } | undefined {
  const at = token.indexOf("@");
  const number = at === -1 ? token : token.slice(0, at);
  const owner = at === -1 ? "" : token.slice(at + 1);
  if (at !== -1 && owner === "") return undefined;
  if (!/^[0-9]+$/.test(number)) return undefined;
  const counter = BigInt(number);
  return counter > MAX_UINT64 ? undefined : { counter, owner };
}
